internal/tofu/context_apply2_test.go

```diff
@@ -4150,3 +4151,265 @@ func TestContext2Apply_excludedModuleRecursive(t *testing.T) {
 <no state>
 	`)
 }
+
+func TestContext2_providerResourceIteration(t *testing.T) {
```

Comment:
Nit: The tests in this and the other "apply" file tend to start with `TestContext2Apply` to differentiate them from the context tests for other phases. (This is _very very occasionally_ useful for using `-run='^TestContext2Apply_'` to filter down the tests by phase, but it's admittedly marginal given how broad a scope these tests cover.)